### PR TITLE
feat: add pointers concept

### PR DESCRIPTION
**Pull request by vaeng**
I would put that concept after "references" and follow up with smart pointers. the trinity of confusion.

I think [windowing system](https://github.com/exercism/javascript/blob/main/exercises/concept/windowing-system/.docs/instructions.md) might be a fitting concept exercise.

### CODE REVIEW

**Review by vaeng — COMMENTED** [review]
Pointers. Haiyaa.
It is difficult to make it short and interesting but not shallow.
Any suggestions?

File: concepts/pointers/introduction.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+# Introduction
```
Comment (vaeng) [inline concepts/pointers/introduction.md:1]:
```suggestion
# Introduction
```

About.md is missing. Please fix before merge.

File: concepts/pointers/.meta/config.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "blurb": "In C++, pointers are variables that store memory addresses. They allow you to directly manipulate memory and access to data structures like arrays and objects.",
```
Comment (siebenschlaefer) [inline concepts/pointers/.meta/config.json:2]: That looks a bit long. Does it fit wherever it is displayed on the website? Can you shorten it?

File: concepts/pointers/introduction.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+Like many other languages, C++ has pointers.
+You already know `references` and pointers are similar, but think of them as a level closer to the inner workings of your computer.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:4]: Here and in the rest of the PR: the use of backticks looks inconsistent. I'd suggest using backticks for code and italics or quotes for technical terms that you want to highlight.

File: concepts/pointers/introduction.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+They're a crucial part of C++, so take some time to really understand them.
+The bare-bone version in this concept is also called _dumb pointer_ or _raw pointer_.
+With modern C++ there are also `smart pointers`, the basic type is not smart at all and you have to handle all the work manually.
+
+Before digging into the details, it's worth understanding the use of pointers.
+Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
+1. When you have large amounts of data, making copies to pass between functions is very inefficient.
+  By passing the memory location of where the data is stored instead, you can dramatically reduce the resource-footprint of our programs.
+2. By passing pointers between functions, you can access and modify the single copy of the data directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:18]:
```suggestion
2. By passing pointers between functions, you can access and modify the single copy of the data directly, meaning that any changes made by one function are immediately visible to other parts of the program.
```

not only then, think threads

File: concepts/pointers/introduction.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+## General Syntax
+
+A pointer declaration in C++ involves specifying the data type the pointer points to, followed by an asterisk (`*`) and the pointer's name.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:22]:
```suggestion
A pointer declaration in C++ involves specifying the data type to which the the pointer is pointing, followed by an asterisk (`*`) and the pointer's name.
```

sounds a little bit better to my (German) ear

File: concepts/pointers/introduction.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+When pointers are declared, they are not automatically initialized.
+Without explicit assignment, a pointer typically holds an indeterminate value, often referred to as a "garbage address."
+While certain compilers might initialize pointers to `nullptr`, this behavior is not guaranteed across all compilers, so it's essential not to rely on it.
+It's best practice to explicitly initialize raw pointers and verify their non-null status before utilization to avoid potential issues.
+
+```cpp
+int* ptr = nullptr; // Declares a pointer and makes sure it is not invalid
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:29]: In the other concepts you've used brace-initialization instead of this assignment-like syntax.

File: concepts/pointers/introduction.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+To assign the address of a variable to a pointer, you use the address-of operator (`&`).
+Dereferencing a pointer is done using the asterisk (`*`) operator.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:33]: The standard call it the "unary `*` operator". They explain that it performs an "indirection" so if you're looking for a name you could call it the "indirection operator".

File: concepts/pointers/introduction.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+Here's a brief example, with a `struct` `Superhero` that has a member variable `superpower`.
+The main function creates a pointer `dianaPrince` to a `Superhero` object (representing Wonder Woman).
+The `->` operator is used to access the member variable `superpower`, showcasing Wonder Woman's iconic "Lasso of Truth."
+
+```cpp
+struct Superhero {
+    std::string superpower;
+};
+
+Superhero* dianaPrince = new Superhero;
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:83]: Do you really want to use `new` before it's introduced in the next section?

File: concepts/pointers/introduction.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+They're a crucial part of C++, so take some time to really understand them.
+The bare-bone version in this concept is also called _dumb pointer_ or _raw pointer_.
+With modern C++ there are also `smart pointers`, the basic type is not smart at all and you have to handle all the work manually.
+
+Before digging into the details, it's worth understanding the use of pointers.
+Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
+1. When you have large amounts of data, making copies to pass between functions is very inefficient.
+  By passing the memory location of where the data is stored instead, you can dramatically reduce the resource-footprint of our programs.
+2. By passing pointers between functions, you can access and modify the single copy of the data directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.
+
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:19]: Points 1 and 2 can also be achieved through references. You could add two more bullet points that explain that pointers can be reassigned to point to a different object, and they can be a null-pointer, i.e. not pointing to any object.

File: concepts/pointers/introduction.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+// Instead of ethan's, the opponent's name stored at the memory address
+std::string passportName = *ethan;
+```
+
+Attention: dereferencing has to be done explicitly, while `references` just worked like an alias.
+
+## Pointer Arithmetic
+
+Pointer arithmetic allows you to perform arithmetic operations on pointers, which is particularly useful when working with arrays.
+Adding an integer to a pointer moves it to a different memory location based on the size of the data type it points to.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:48]: I think this sentence can be hard to understand for beginners. How about something like "makes it point to a different element"?

**Review by siebenschlaefer — COMMENTED** [review]
Not bad!

This whole documents describes pointers as "memory addresses".  That is not wrong, at least on a practical level.  

Formally (I think) neither the [C++17 standard](https://wg21.link/n4659) nor the [C11 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) (which is part of the C++17 standard) define pointers in terms of "memory" but rather with "addresses" and "objects". Instead of defining what a pointer is they describe how you can derive it, when it is valid, how you can use it.
You could consider replacing many occurrences of "memory" in this PR with "objects", e.g. instead of "Pointers are a way to share memory" you could write "Pointers are a way to share objects". That would match the official terminology.

If you want to describe pointers as addresses in memory that's fine with me, just keep in mind that this is not in the official specification, and you cannot refer to some formal definition but have to describe it in your own words.

I'm missing a warning in the section about dynamic memory allocation that modern C++ code should use smart pointers (perhaps with a link to the other concept). I'd rather remove that whole section and put it into a different concept about dynamic memory allocation.

File: concepts/pointers/.meta/config.json (hunk 2 above)
Comment (vaeng) [reply concepts/pointers/.meta/config.json:2]: It's less than 180 chars, I think that is still in the okay range?

> "blurb: A short description of this concept. Its length must be <= 350. Markdown is not supported (required)"
From: https://exercism.org/docs/building/tracks/concepts#h-file-meta-config-json

File: concepts/pointers/introduction.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+## Pointer Arithmetic
+
+_Pointer arithmetic_ allows you to perform arithmetic operations on pointers, which is particularly useful when working with arrays.
+Adding an integer to a pointer makes it point to a different element.
+
+```cpp
+// Stargate addresses
+int gateAddresses[] = {462, 753, 218, 611, 977};
+// 'ptr' points to the first element of 'gateAddresses'
+int* ptr{gateAddresses};
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:54]: To understand this line one has to know about how arrays decay to pointers.  
I'd suggest `&gateAdresses[0]` instead of `gateAddresses`.  
That doesn't change anything else but I think it's easier to understand.

File: concepts/pointers/introduction.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+Attention: dereferencing has to be done explicitly, while _references_ just worked like an alias.
+
+## Pointer Arithmetic
+
+_Pointer arithmetic_ allows you to perform arithmetic operations on pointers, which is particularly useful when working with arrays.
+Adding an integer to a pointer makes it point to a different element.
+
+```cpp
+// Stargate addresses
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:51]: Since this concept is about pointers and addresses using the term "stargate addresses" might be a little bit confusing (one is the technical term related to the concept, the other is the fictional term for interstellar travel.)

File: concepts/pointers/introduction.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+Undefined behavior can manifest in unexpected program outcomes, crashes, or even security vulnerabilities.
+One infamous example of the consequences of undefined behavior occurred in the [explosion of the Ariane 5 rocket][ariane-flight-v88] in 1996, where a software exception caused by the conversion of a 64-bit floating-point number to a 16-bit signed integer led to a catastrophic failure.
+~~~~
+
+## Dynamic Memory Allocation and Null Pointers
+
+In C++, `new` is used to dynamically allocate memory for objects on the heap during runtime.
+It returns a pointer to the allocated memory address, allowing objects to have lifetimes not tied to a specific scope.
+However, it's important to always pair `new` with `delete` to properly deallocate memory.
+Dynamic memory allocation should be used cautiously to prevent common issues like memory leaks and accessing deallocated memory.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:72]: I would even use stronger language to discourage the use of "naked" or "raw" `new`.

File: concepts/pointers/introduction.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+A pointer should either hold the address of a valid object, or be set to `nullptr`.
+That way you only need to test pointers for null, and can assume any non-null pointer is valid.
+When an object is destroyed, any pointers to the destroyed object will be left _dangling_ (they will not be automatically set to `nullptr`).
+It is your responsibility to detect these cases and ensure those pointers are subsequently set to `nullptr`.
+
+~~~~exercism/advanced
+In older code, you might encounter two alternatives to `nullptr`.
+Firstly, the literal `0` is specifically interpreted as a null value for pointers, though it's the only scenario where an integral literal can be assigned to a pointer.
+Secondly, the `preprocessor macro` `NULL`, inherited from C and defined in the `<cstddef>` header, is another representation of a null pointer, though its usage is less common in modern C++ code.
```
Comment (siebenschlaefer) [inline concepts/pointers/introduction.md:96]: Is the term "preprocessor macro" in backticks on purpose?

**Review by siebenschlaefer — APPROVED** [review]
Looks good.

My new suggestions are mostly nitpicks, feel free to ignore or disregard them.

**Comment by vaeng** [comment]
I moved the section about dynamic memory allocation to the `about.md` and marked it as advanced.
